cmd/agent/app/flags.go

```diff
@@ -24,54 +24,71 @@ import (
 	"flag"
 	"fmt"
 	"strings"
+	"github.com/spf13/viper"
 )
 
+type stringSliceFlag struct {
+	slice *[]string
+}
+
 // Bind binds the agent builder to command line options
-func (b *Builder) Bind(flags *flag.FlagSet) {
+func (b *Builder) GenerateFlags() *flag.FlagSet {
+	flagSet := &flag.FlagSet{}
+
 	for i := range b.Processors {
 		p := &b.Processors[i]
 		name := "processor." + string(p.Model) + "-" + string(p.Protocol) + "."
-		flags.IntVar(
-			&p.Workers,
+		flagSet.Int(
 			name+"workers",
 			p.Workers,
```

Comment:
Yes it is.

`GenerateFlags` now generates only flag set which is added to given cobra command before executing it. After `command.Execute` properties are obtained from viper via `builder.Parse` or `NewProperties()`.

cobra is used only for command line parameters. Viper can consume cobra configuration and add other features like binding to env variables or config files. However viper does not support pointers e.g. `viper.GetStringVar(var, 'fooFlag')`, therefore I have split flags generation and binding to objects. (Viper also supports watching configuration files changes which emits event that something has changed https://github.com/spf13/viper#watching-and-re-reading-config-files, from design perspective I think it's better to have method like `newProperties()`  it can be used to compate what has changed etc.)